### ISSUE TITLE
Incorrect inheritance of method type

### ISSUE THREAD

**Issue by Perfectoff** [microsoft/TypeScript#60512]
### 🔎 Search Terms

Incorrect inheritance of method type

### 🕗 Version & Regression Information

any version

### ⏯ Playground Link

https://www.typescriptlang.org/play/?ts=5.6.3#code/FDAE9BLA7AXBTATgMwIYGN6gIKgN5gREDm8sAaqgDYAUAJqrKqAFz4C+AlK6NAK4BbAEZIA3IXDsQRdFVQBneaABCUAQAcq8AfDhLcBIiTKVaDJjzyJ4qOgHtoVAJ6gAbtTb9hSLvlAB6fyhodDtEa3RYUFgndSw7ZFAAcnNUJIA6UABJWCUdb0Rk9yokqCVoOyi6eFlUazpg6IALLCEFLB1YJrs6CSMia1g+RGhQVPTi8X7JPqk+rSihHlwAXl54AHcVGk4piAW3ah4vEUK1oXTSCmoaPC498GL02DsAMUgAD3g6HdEAoOIlVAzEQfDgkB0oCQiDCbFAAGFUNAKlFrLZQOoYXFELBIPAlAlQGDqsgYN9QDQ0XQYMRki93l86ElOCAgA

### 💻 Code

```ts
    interface A {
        getVal(data : {}) : number;
    }

    class B implements A {
        getVal(data : {readonly val: number}) {  // incorrect type of 'data'. Its member 'val' is not declared in the base method
            return data.val;
        }
    }

    let b : A = new B();
    let val : number = b.getVal({});
    val.toFixed();  // got a runtime error:  Cannot read properties of undefined (reading 'toFixed')
```

### 🙁 Actual behavior

In this code, class 'A' implements a method with an incorrect argument type, which is incompatible with the original interface. However, TypeScript does not report any errors. As a result, we get a runtime error.

### 🙂 Expected behavior

Should be an error in the implementation

**Comment by MartinJohns** [comment]
This is working as intended. See the FAQ: https://github.com/microsoft/TypeScript/wiki/FAQ#why-method-bivariance

**Comment by whzx5byb** [comment]
It's recommended to always use "function property" instead of "method" in your interfaces if you don't like the bivariance behavior.
```diff
interface A {
-    getVal(data : {}) : number;
+    getVal: (data: {}) => number;
}
```

**Comment by Perfectoff** [comment]
> It's recommended to always use "function property" instead of "method" in your interfaces if you don't like the bivariance behavior.
> 
> interface A {
> -    getVal(data : {}) : number;
> +    getVal: (data: {}) => number;
> }

But what is the reason for such different behavior? The "function property" is essentially the same as a method in terms of typing. Therefore, this can be considered a bug in the typing of methods.

**Comment by Perfectoff** [comment]
> This is working as intended. See the FAQ: https://github.com/microsoft/TypeScript/wiki/FAQ#why-method-bivariance

That doesn't apply to my case. There, the reason is in the combination of types, leading to such casting of functions:

`(arg: number|string)=>any`  casts to  `(arg: number|string|boolean)=>any` .  And this is correct.

But in my case:  `(arg: {val:number})=>any`  casts  to `(arg :{})=>any` .  And this is incorrect!

**Comment by Perfectoff** [comment]
>  `(arg: number|string)=>any` casts to `(arg: number|string|boolean)=>any` . And this is correct.
> 
Oh, sorry, i was wrong, there is the opposite casting

**Comment by Perfectoff** [comment]
> It's recommended to always use "function property" instead of "method" in your interfaces if you don't like the bivariance behavior.

By the way, this is not just about interfaces, but about overriding methods too.  So that means we should never use methods in classes.  And that's not a bug, that's a feature... Great

**Comment by MartinJohns** [comment]
> But what is the reason for such different behavior?

It's explained in the link I posted. I suggest you to read it.

**Comment by Perfectoff** [comment]
> > But what is the reason for such different behavior?
> 
> It's explained in the link I posted. I suggest you to read it.

I meant difference between "function property" and "method".  I didn't find any explanation about this in your link.

Why is there bivariance here:
```ts 
interface A {
    getVal(data : {}) : number;
}
```
and there is no bivariance here:
```ts
interface A {
    getVal: (data: {}) => number;
}
```

**Comment by jcalz** [comment]
#18654

**Comment by Perfectoff** [comment]
I just can't understand why TypeScript can't report an error in my particular case? Everything is obvious and clear there. We have a method declaration and its implementation, which is incompatible with the declaration. What could be the problem with the check there?

And the link that was given earlier described more complex cases, with generics and type unions. Okay, these cases can be ambiguous, so TypeScript could skip them. But in my example there is nothing like that, everything is simple and clear. Why not fix such obvious errors?

**Comment by MartinJohns** [comment]
Because methods are bivariant, it's not an error but the intended design. And your `{readonly val: number}` type is bivariant compatible with the type `{}`. If you want covariant behavior you can use function properties.

**Comment by Perfectoff** [comment]
> Because methods are bivariant, it's not an error but the intended design. And your `{readonly val: number}` type is bivariant compatible with the type `{}`. If you want covariant behavior you can use function properties. The FAQ link and the linked GitHub issue by jcalz talk about this.

According to the [explanation](https://github.com/microsoft/TypeScript/wiki/FAQ#why-method-bivariance), this does not seem like an intended design. Rather, it seems like a problem that has not yet been solved. It says that enabling this check will show a large number of errors in old codes that relied on bivariance.
I suppose the same thing once happened with function bivariance, when users asked to add the --strictFunctionTypes option. Apparently, there was a similarly large pile of old code that relied on function bivariance. But the option was still added. Progress cannot be stopped, right? :wink:

Personally, I think using this bivariance is bad practice. Those examples shown in the link do not seem to me to be good programming design. Rather, they are some kind of patches. I would like to see a more real-life example of code that requires bivariance and cannot do otherwise. To understand for what sake we sacrifice code safety.

**Comment by jcalz** [comment]
https://github.com/microsoft/TypeScript/issues/49432#issuecomment-1151209921

Welcome to the TypeScript [Unsoundness](https://www.typescriptlang.org/docs/handbook/type-compatibility.html#a-note-on-soundness) Support Group! I'm [@jcalz](https://github.com/jcalz) and I have been a member for six years. Be careful with the folding chairs, they've been known to collapse. Today we'll continue our discussion of how the TypeScript Language Design [Non-Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals#non-goals) interact with the stages of grief.

**Comment by Perfectoff** [comment]
Well, I found a workaround for this so-called "intended design".

```ts
    interface A {
        getVal<T>(data : {}&T) : number;
        //or  
        //getVal<T extends {}>(data : T) : number;
    }

    class B implements A {
        getVal(data : {readonly val: number}) { // TS error
            return data.val;
        }
    }
    
    class C implements A {
        getVal(data : {}) {  // ok 
            return 0;
        }
    }
```
